cpp/include/cudf/detail/iterator.cuh

```diff
@@ -0,0 +1,118 @@
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <thrust/iterator/iterator_adaptor.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+namespace cudf {
+namespace experimental {
+
+/** -------------------------------------------------------------------------*
+ * @brief value accessor of column with null bitmask
+ * A unary functor returns scalar value at `id`.
+ * `operator() (cudf::size_type id)` computes `element` and valid flag at `id`
+ *
+ * the return value for element `i` will return `column[i]`
+ * if it is valid, or `null_replacement` if it is null.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+template <typename Element>
+struct value_accessor
+{
+  column_device_view const col;         ///< column view of column in device
+  Element const null_replacement{};     ///< value returned when element is null
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * @param[in] null_replacement The value to return for null elements
+ * -------------------------------------------------------------------------**/
+  value_accessor(column_device_view const _col, Element null_val)
+    : col{_col}, null_replacement{null_val}
+  {
+    // verify valid is non-null, otherwise, is_valid() will crash
+    CUDF_EXPECTS(_col.null_mask() != nullptr, "non-null bit mask is required");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  Element operator()(cudf::size_type i) const {
+    return col.is_valid_nocheck(i) ?
+      col.element<Element>(i) : null_replacement;
+  }
+};
+} //namespace experimental
+
+
+/**
+ * @brief Constructs an iterator over a column's values that replaces null
+ * elements with a specified value.
+ *
+ * Dereferencing the returned iterator for element `i` will return `column[i]`
+ * if it is valid, or `null_replacement` if it is null.
+ *
+ * @tparam Element The type of elements in the column
+ * @param column The column to iterate
+ * @param null_replacement The value to return for null elements
+ * @return auto Iterator that returns valid column elements, or a null
+ * replacement value for null elements.
+ */
+template <typename Element>
+auto make_null_replacement_iterator(column_device_view const column,
+                                    Element const null_replacement = Element{0})
+{
+  return thrust::make_transform_iterator(
+      thrust::counting_iterator<cudf::size_type>{0},
+      experimental::value_accessor<Element>{column, null_replacement});
+}
+
+template <typename Element,
+          typename Iterator_Index = thrust::counting_iterator<cudf::size_type>>
+auto make_null_replacement_custom_index_iterator(
```

Comment:
What's the usecase for this?

Reply:
For gather and scatter, this could be useful
This feature was present in old iterator, so i thought, it could be useful. It is not used anywhere now.

Reply:
documentation updated.

Reply:
custom_indexed_iterator removed from PR